aptos-move/block-executor/src/executor.rs

```diff
@@ -199,6 +210,50 @@ where
         }
     }
 
+    fn commit_hook(
+        &self,
+        txn_idx: TxnIndex,
+        versioned_cache: &MVHashMap<T::Key, T::Value, ExecutableTestType>,
+        last_input_output: &TxnLastInputOutput<T::Key, E::Output, E::Error>,
+        base_view: &S,
+    ) {
+        let (num_deltas, delta_keys) = last_input_output.delta_keys(txn_idx);
+        let mut delta_writes = Vec::with_capacity(num_deltas);
+        for k in delta_keys {
+            // Note that delta materialization happens concurrenty, but under concurrent
+            // commit_hooks (which may be dispatched by the coordinator), threads may end up
+            // contending on delta materialization of the same aggregator. However, the
+            // materialization is based on previously materialized values and should not
+            // introduce long critical sections. Moreover, with more aggregators, and given
+            // that the commit_hook will be performed at dispersed times based on the
+            // completion of the respetive previous tasks of threads, this should not be
+            // an immediate bottleneck - confirmed by an experiment with 32 core and a
+            // single materialized aggregator. If needed, the contention may be further
+            // mitigated by batching consecutive commit_hooks.
+            let committed_delta = versioned_cache
+                .materialize_delta(&k, txn_idx)
+                .unwrap_or_else(|| {
+                    let storage_value = base_view
+                        .get_state_value_bytes(&k)
+                        .expect("No base value for committed delta in storage")
+                        .map(|bytes| deserialize(&bytes))
+                        .expect("Cannot deserialize base value for committed delta");
+
+                    versioned_cache.set_aggregator_base_value(&k, storage_value);
+                    versioned_cache
+                        .materialize_delta(&k, txn_idx)
```

Comment:
Why not make materialization return Result and wrap Ok(u128), Err(DeltaOp) ? Then we do not need to re-materialize (which does a read and a delta traversal), and can apply delta instead?